In [2]:
import numpy as np

### 사과 2개, 귤 3개
### 사과: 100원, 귤: 150원
### 소비세: 10%
### 지불금액 ??

In [6]:
# 곱셈 계층

class MulLayer:
    
    def __init__(self):
        self.x = None
        self.y = None
    
    def forward(self, x, y):
        self.x = x
        self.y = y
        out = x * y
        return out

    def backward(self, dout):
        dx = dout * self.y
        dy = dout * self.x
        
        return dx, dy

In [7]:
apple = 100
apple_num = 2
tax = 1.1

mul_apple_layer = MulLayer()
mul_tax_layer = MulLayer()

# 순전파
apple_price = mul_apple_layer.forward(apple, apple_num)
price = mul_tax_layer.forward(apple_price, tax)

print(price)

220.00000000000003


In [8]:
# 덧셈 계층

class AddLayer:
    def __int__(self):
        pass
    
    def forward(self, x, y):
        out = x + y
        return out
    
    def backward(self, dout):
        dx = dout * 1
        dy = dout * 1
        return dx, dy

In [10]:
apple = 100
apple_num = 2
orange = 150
orange_num = 3
tax = 1.1

# (apple * apple_num + orange * orange_num) * tax = price

# layer 설정
mul_apple_layer = MulLayer()
mul_orange_layer = MulLayer()
add_apple_orange_layer = AddLayer()
mul_tax_layer = MulLayer()

# 순전파

# apple_price = apple * apple_num
apple_price = mul_apple_layer.forward(apple, apple_num)
# orange_price = orange * orange_num
orange_price = mul_orange_layer.forward(orange, orange_num)
# all_price = apple_price + orange_price
all_price = add_apple_orange_layer.forward(apple_price, orange_price)
# price = tax * all_price
price = mul_tax_layer.forward(all_price, tax)

# 역전파
dprice = 1
dall_price, dtax = mul_tax_layer.backward(dprice)
dapple_price, dorange_price = add_apple_orange_layer.backward(dall_price)
dorange, dorange_num = mul_orange_layer.backward(dorange_price)
dapple, dapple_num = mul_apple_layer.backward(dapple_price)

print("Price:", int(price))
print("dApple:", dapple)
print("dApple_num:", int(dapple_num))
print("dOrange:", dorange)
print("dOrange_num:", int(dorange_num))
print("dTax:", dtax)

Price: 715
dApple: 2.2
dApple_num: 110
dOrange: 3.3000000000000003
dOrange_num: 165
dTax: 650


### dTax: 부과세가 1만큼 증가하면 최종 가격은 650원 증가
### dApple_num: 사과 개수가 1개 증가하면 최종 가격은 110원 증가
### dOrange_num: 오렌지 개수가 1개 증가하면 최종 가격은 165원 증가